src/core/AudioEngine.h

```diff
@@ -284,11 +296,65 @@ class AudioEngine : public H2Core::Object
 	AudioEngine();
 };
 
+
+/**
+ * AudioEngineLocking
+ *
+ * This is a base class for shared data structures which may be
+ * modified by the AudioEngine. These should only be modified or
+ * trusted by a thread holding the AudioEngine lock.
+ *
+ * Any class which implements a data structure which can be modified
+ * by the AudioEngine can inherit from this, and use the protected
+ * "assertLocked()" method to ensure that methods are called only
+ * with appropriate locking.
+ *
+ * Checking is only done on debug builds.
+ */
+class AudioEngineLocking {
+
+	bool m_bNeedsLock;
+
+protected:
+	/**
+	 *  Assert that the AudioEngine lock is held if needed.
+	 */
+	void assertLocked() const {
+#ifndef NDEBUG
+		if ( m_bNeedsLock ) {
+			AudioEngine::get_instance()->assertLocked();
+		}
+#endif
+	}
+
+
+public:
+
+	/**
+	 * The audio processing thread can modify some PatternLists. For
+	 * these structures, the audio engine lock must be held for any
+	 * thread to access them.
+	 */
+	void setNeedsLock( bool bNeedsLock ) {
+		m_bNeedsLock = bNeedsLock;
+	}
+
+	AudioEngineLocking() {
+		m_bNeedsLock = false;
+	}
+};
+
+
 inline float AudioEngine::getElapsedTime() const {
 	return m_fElapsedTime;
 }
-	
-};
 
+inline void AudioEngine::assertLocked( ) {
```

Comment:
What about renaming this to `assertAudioEngineLocked()` ? It sounds overflous in this class definition, but when used in a derived class, it might make the intention about what is beeing locked more clear

Reply:
Hmm. Yes, agreed. It's probably also good to have the method name different from that in AudioEngine.